### ISSUE TITLE
Custom Datatypes in Tensors

### ISSUE THREAD

**Issue by SharonBarak** [pytorch/pytorch#45051]
## 🚀 Feature

The ability to create tensors with custom objects.

## Motivation

I'm working on machine learning with homomorphic encrypted data, which is stored in a Python class with operator overloading. Currently tensor attributes can be one of a few primitive types, and I am not able to create a Pytorch tensor which holds this data.

## Pitch

Numpy is able to create ndarrays holding custom Python objects and calls the overloaded operators for its functions. Something similar here would be great.

## Alternatives

I am able to create a neural network from scratch but it would be nice to have compatibility with Pytorch.

**Comment by mruberry** [comment]
Thanks for the suggestion, @SharonBarak.

**Comment by rgommers** [comment]
numpy `object` dtype arrays can be quite useful, but it's hard to see how that translates to GPU or other devices.

**Comment by SharonBarak** [comment]
Hi @rgommers and @mruberry , just wanted to check in and see if there has been any progress on this suggestion.

**Comment by rgommers** [comment]
Not that I know of. The "needs research" label basically means "unclear what to do". Adding more detail would be useful. Like, does it actually need to be a Python-level object? That is unlikely to be accepted, because it cannot be implemented on a GPU. So why use PyTorch over NumPy here?

**Comment by SharonBarak** [comment]
Since PyTorch offers a lot of functionality for working with machine learning, it would be nice to be able to use it. Numpy does support creating ndarrays with general Python objects but it doesn't have all the functionality that comes with PyTorch.

Would creating Tensors that hold general Python-level objects be too difficult, even if it is only implemented on the CPU?

**Comment by rgommers** [comment]
How difficult it is isn't really the right question - I don't think there's much of a desire to add CPU-only functionality. GPU (and other device) support is an essential part of PyTorch, and on CPU it's not necessarily faster or more flexible than NumPy. So why not just point people to NumPy?

> Numpy does support creating ndarrays with general Python objects but it doesn't have all the functionality that comes with PyTorch.

Well, they're just different libraries. NumPy works natively with SciPy, Scikit-learn and many other libraries, so I think functionality is much broader than PyTorch. 

> Numpy is able to create ndarrays holding custom Python objects and calls the overloaded operators for its functions. Something similar here would be great.

You can already create a Tensor-like class and use the rest of the PyTorch API (including `torch.nn`, which is probably what you're after mostly). Can you see if https://pytorch.org/docs/master/notes/extending.html#extending-torch serves your needs? You can look at https://github.com/pytorch/nestedtensor for a nice example.
